src/java/org/apache/cassandra/tcm/AtomicLongBackedProcessor.java

```diff
@@ -177,17 +177,17 @@ public synchronized EntryHolder getEntries(Epoch since, Epoch until)
         public LogState getLogState(Epoch start, Epoch end)
         {
             EntryHolder state = getEntries(Epoch.EMPTY);
-            ClusterMetadata metadata = new ClusterMetadata(DatabaseDescriptor.getPartitioner());;
+            ClusterMetadata metadata = new ClusterMetadata(DatabaseDescriptor.getPartitioner());
             Iterator<Entry> iter = state.iterator();
             ImmutableList.Builder<Entry> rest = new ImmutableList.Builder<>();
             while (iter.hasNext())
             {
                 Entry current = iter.next();
                 if (current.epoch.isAfter(end))
                     break;
-                if (current.epoch.isEqualOrBefore(start))
+                if (current.epoch.isDirectlyAfter(metadata.epoch))
```

Comment:
`metadata.epoch` will always be 0 here right? So what happens if start is not 0? How can a transformation that doesn't start from the beginning work on an empty epoch when the transformation is supposed to operate on one that is not empty?

Reply:
You are right; I have reverted to the previous logic to build snapshot _up to_ start epoch; however I have added the invariant now `Invariants.checkState(current.epoch.isDirectlyAfter(metadata.epoch));` just to make sure.

My logic was that we are executing this in a loop, and we want to apply only consecutive epochs. I think previous version of the code only worked because we were returning all values as log entries rather than in the snapshot.